numpy/_core/include/numpy/ndarraytypes.h

```diff
@@ -641,6 +646,55 @@ typedef struct _PyArray_Descr {
 
 } PyArray_Descr;
 
+#else  /* internal build */
+
+// TODO: This split definition only exists for piece-meal transitioning
+//       as it allows change internal use without worrying about public API.
+typedef struct _PyArray_Descr {
+        PyObject_HEAD
+        PyTypeObject *typeobj;
+        char kind;
+        char type;
+        char byteorder;
+        char flags;
+        int type_num;
+        int elsize;
+        int alignment;
+        /* except hash, the below fields will be legacy descriptor specific */
+        struct _arr_descr *unreachable_subarray;
+        PyObject *unreachable_fields;
+        PyObject *unreachable_names;
+        PyArray_ArrFuncs *f;
+        PyObject *unreachable_metadata;
+        NpyAuxData *unreachable_c_metadata;
+        npy_hash_t hash;
+} PyArray_Descr;
+
+#endif  /* internal build */
+
+
+/*
+ * Semi-private struct with additional field of legacy descriptors (must
+ * check NPY_DT_is_legacy before casting/accessing).
```

Comment:
A note here for confused code searchers saying this is for dealing with structured arrays and legacy dtype attributes internally and the long-term plan is to minimize its use in favor of the real struct.